src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/ExtractToNewComponentCodeActionResolver.cs

```diff
@@ -0,0 +1,146 @@
+using Microsoft.AspNetCore.Razor.LanguageServer.Hosting;
+using Microsoft.AspNetCore.Razor.Utilities;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor.Protocol.CodeActions;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+using Microsoft.CodeAnalysis.Razor.Protocol;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Newtonsoft.Json.Linq;
+using System.Globalization;
+using System.Text.Json;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.CodeActions.Razor;
+internal sealed class ExtractToNewComponentCodeActionResolver : IRazorCodeActionResolver
+{
+    private static readonly Workspace s_workspace = new AdhocWorkspace();
+
+    private readonly IDocumentContextFactory _documentContextFactory;
+    private readonly LanguageServerFeatureOptions _languageServerFeatureOptions;
+    private readonly IClientConnection _clientConnection;
+    public ExtractToNewComponentCodeActionResolver(
+        IDocumentContextFactory documentContextFactory,
+        LanguageServerFeatureOptions languageServerFeatureOptions,
+        IClientConnection clientConnection)
+    {
+        _documentContextFactory = documentContextFactory ?? throw new ArgumentNullException(nameof(documentContextFactory));
+        _languageServerFeatureOptions = languageServerFeatureOptions ?? throw new ArgumentNullException(nameof(languageServerFeatureOptions));
+        _clientConnection = clientConnection ?? throw new ArgumentNullException(nameof(clientConnection));
+    }
+
+    public string Action => LanguageServerConstants.CodeActions.ExtractToNewComponentAction;
+
+    public async Task<WorkspaceEdit?> ResolveAsync(JsonElement data, CancellationToken cancellationToken)
+    {
+        if (data.ValueKind == JsonValueKind.Undefined)
+        {
+            return null;
+        }
+
+        var actionParams = JsonSerializer.Deserialize<ExtractToNewComponentCodeActionParams>(data.GetRawText());
+        if (actionParams is null)
+        {
+            return null;
+        }
+
+        if (!_documentContextFactory.TryCreate(actionParams.Uri, out var documentContext))
+        {
+            return null;
+        }
+
+        var componentDocument = await documentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+        if (componentDocument.IsUnsupported())
+        {
+            return null;
+        }
+
+        if (!FileKinds.IsComponent(componentDocument.GetFileKind()))
+        {
+            return null;
+        }
+
+        var path = FilePathNormalizer.Normalize(actionParams.Uri.GetAbsoluteOrUNCPath());
+        var directoryName = Path.GetDirectoryName(path) ?? throw new InvalidOperationException($"Unable to determine the directory name for the path: {path}");
+        var templatePath = Path.Combine(directoryName, "Component");
+        var componentPath = FileUtilities.GenerateUniquePath(templatePath, ".razor");
```

Comment:
I'm surprised that calling `Path` functions on a normalized path successfully. It worries me that we'll hit an issue and it's entirely possible that the final path will have fixed slashes on Windows. cc @davidwengier for his thoughts here.

Reply:
Do you mean "mixed" slashes on Windows? I've certainly seen that in other places and it hasn't caused problems, but might be a little riskier here I guess as this is turned in to a Uri and sent back to the client. Maybe the UriBuilder fixes things?

I think there is a good argument that a bunch of this sort of code, which is present in some form here and in the other "Extract" code action, can be moved to a helper class, shared, and get some tests written. To be clear though Hector, that doesn't need to happen in this PR :)

Reply:
> To be clear though Hector, that doesn't need to happen in this PR :)

Correct! I recognize that this sort of code already exists, and I think it's likely problematic. We should definitely do as you suggest @davidwengier, but also as you say, not in this PR. :smile: